c++/src/kj/io.h

```diff
@@ -78,13 +78,22 @@ class OutputStream {
 public:
   virtual ~OutputStream() noexcept(false);
 
-  virtual void write(const void* buffer, size_t size) = 0;
-  // Always writes the full size.  Throws exception on error.
-
-  virtual void write(ArrayPtr<const ArrayPtr<const byte>> pieces);
-  // Equivalent to write()ing each byte array in sequence, which is what the default implementation
-  // does.  Override if you can do something better, e.g. use writev() to do the write in a single
-  // syscall.
+  virtual void write(ArrayPtr<const byte> data, 
```

Comment:
I think this signature for write() didn't work out well after all. Sorry for sending you down the wrong path.

In many places, you've ended up implementing this in terms of a `writeImpl` that writes a single buffer. Meanwhile, most callers only pass a single buffer. So it actually turns out to just be a lot of wasted checks. Also in a lot of these cases you've declared `write()` inline in the header file, but this is inappropriate since a virtual call cannot be inlined, and the `forEachPiece()` invocation is too complex to online anyway, so these method bodies need to be moved to the source files to avoid bloating build times.

I think we should go back to having two methods, one that takes `ArrayPtr<const byte>` and one that takes `ArrayPtr<const ArrayPtr<const byte>>`, and give the latter a default implementation that loops and invokes the former.